File: map_generator.py
# map_generator.py
import random

MAP_SIZE = 5

def count_turns(path):
    """計算這條路徑總共轉彎了幾次"""
    if len(path) < 3:
        return 0
    turns = 0
    for i in range(1, len(path) - 1):
        prev_pt = path[i-1]
        curr_pt = path[i]
        next_pt = path[i+1]
        
        dir1_x, dir1_y = curr_pt[0] - prev_pt[0], curr_pt[1] - prev_pt[1]
        dir2_x, dir2_y = next_pt[0] - curr_pt[0], next_pt[1] - curr_pt[1]
        
        if (dir1_x, dir1_y) != (dir2_x, dir2_y):
            turns += 1
    return turns

def generate_strict_path():
    """生成『絕對唯一通路』『絕無捷徑』且『至少三轉彎』的迷宮"""
    
    # 邊界起點名單
    border_positions = []
    for i in range(MAP_SIZE):
        border_positions.append((i, 0))
        border_positions.append((i, MAP_SIZE - 1))
        if i != 0 and i != MAP_SIZE - 1:
            border_positions.append((0, i))
            border_positions.append((MAP_SIZE - 1, i))
            
    while True:
        start_x, start_y = random.choice(border_positions)
        path = [(start_x, start_y)]
        visited = set(path)
        
        curr_x, curr_y = start_x, start_y
        stuck = False
        
        # 預期長度：既然不能有捷徑，在5x5裡通常會走 7~10 格
        target_length = random.randint(7, 10)
        
        for _ in range(target_length):
            possible_moves = []
            
            for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                nx, ny = curr_x + dx, curr_y + dy
                
                # 基本檢查：必須在大樓內，且沒走過
                if 0 <= nx < MAP_SIZE and 0 <= ny < MAP_SIZE and (nx, ny) not in visited:
                    
                    # ⭐ 關鍵防捷徑檢查：這格 nx, ny 的「四周」
                    # 除了牠走過來的上一格 (curr_x, curr_y) 之外，絕對不能碰觸到任何其他已經是 1 的路徑格！
                    adjacent_path_count = 0
                    for adj_dx, adj_dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                        ax, ay = nx + adj_dx, ny + adj_dy
                        if (ax, ay) in visited:
                            adjacent_path_count += 1
                    
                    # 如果四周碰到的舊路徑格 > 1 (代表除了來時路，還碰到了別的舊路)，走過去就會變捷徑！
                    # 所以只有當鄰近舊路徑只有1個時，這格才安全，牆壁才不會破
                    if adjacent_path_count <= 1:
                        possible_moves.append((nx, ny))
            
            if not possible_moves:
                stuck = True
                break # 撞進死胡同（被自己堵死），重來
                
            curr_x, curr_y = random.choice(possible_moves)
            path.append((curr_x, curr_y))
            visited.add((curr_x, curr_y))
            
        if stuck:
            continue
            
        # 檢查轉彎次數是否符合標準 (>=3)
        if count_turns(path) >= 3:
            end_pos = path[-1]
            
            # 生成矩陣地圖
            dungeon_map = [[0 for _ in range(MAP_SIZE)] for _ in range(MAP_SIZE)]
            for x, y in path:
                dungeon_map[y][x] = 1
                
            return dungeon_map, (start_x, start_y), end_pos, path

File: map_generator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAP_SIZE 5

typedef struct {
    int x;
    int y;
} Point;

typedef struct {
    int count;
    Point* points;
} Path;

// 計算路徑中的轉彎次數
int count_turns(Path* path) {
    if (path->count < 3) return 0;
    
    int turns = 0;
    for (int i = 1; i < path->count - 1; i++) {
        int dir1_x = path->points[i].x - path->points[i-1].x;
        int dir1_y = path->points[i].y - path->points[i-1].y;
        int dir2_x = path->points[i+1].x - path->points[i].x;
        int dir2_y = path->points[i+1].y - path->points[i].y;
        
        if (dir1_x != dir2_x || dir1_y != dir2_y) {
            turns++;
        }
    }
    return turns;
}

// 核心演算法：生成絕對無捷徑、唯一通路的單向軌道
Path* generate_strict_path() {
    static int seeded = 0;
    if (!seeded) {
        srand((unsigned int)time(NULL));
        seeded = 1;
    }
    
    while (1) {
        int start_x, start_y;
        int border_idx = rand() % 16; 
        if (border_idx < 5) {          
            start_x = border_idx; start_y = 0;
        } else if (border_idx < 10) {   
            start_x = border_idx - 5; start_y = MAP_SIZE - 1;
        } else if (border_idx < 13) {   
            start_x = 0; start_y = border_idx - 9;
        } else {                        
            start_x = MAP_SIZE - 1; start_y = border_idx - 12;
        }
        
        Path* path = (Path*)malloc(sizeof(Path));
        path->points = (Point*)malloc(sizeof(Point) * 100);
        path->count = 0;
        
        int* visited = (int*)malloc(sizeof(int) * MAP_SIZE * MAP_SIZE);
        memset(visited, 0, sizeof(int) * MAP_SIZE * MAP_SIZE);
        
        path->points[0].x = start_x;
        path->points[0].y = start_y;
        path->count = 1;
        visited[start_y * MAP_SIZE + start_x] = 1;
        
        int curr_x = start_x;
        int curr_y = start_y;
        int stuck = 0;
        
        int target_length = 7 + rand() % 4; 
        
        // 🌟 ✅ 修正 1：將條件改為 step <= target_length，確保走滿完整的總步數，與 Python 完美對齊！
        for (int step = 1; step <= target_length; step++) {
            Point possible_moves[4];
            int move_count = 0;
            
            int dx_list[] = {-1, 1, 0, 0};
            int dy_list[] = {0, 0, -1, 1};
            
            for (int d = 0; d < 4; d++) {
                int nx = curr_x + dx_list[d];
                int ny = curr_y + dy_list[d];
                
                if (nx >= 0 && nx < MAP_SIZE && ny >= 0 && ny < MAP_SIZE && !visited[ny * MAP_SIZE + nx]) {
                    int adjacent_path_count = 0;
                    for (int a = 0; a < 4; a++) {
                        int ax = nx + dx_list[a];
                        int ay = ny + dy_list[a];
                        if (ax >= 0 && ax < MAP_SIZE && ay >= 0 && ay < MAP_SIZE) {
                            if (visited[ay * MAP_SIZE + ax]) {
                                adjacent_path_count++;
                            }
                        }
                    }
                    
                    if (adjacent_path_count <= 1) {
                        possible_moves[move_count].x = nx;
                        possible_moves[move_count].y = ny;
                        move_count++;
                    }
                }
            }
            
            if (move_count == 0) {
                stuck = 1;
                free(visited);
                free(path->points);
                free(path);
                break;
            }
            
            int chosen = rand() % move_count;
            curr_x = possible_moves[chosen].x;
            curr_y = possible_moves[chosen].y;
            path->points[path->count].x = curr_x;
            path->points[path->count].y = curr_y;
            path->count++;
            visited[curr_y * MAP_SIZE + curr_x] = 1;
        }
        
        if (stuck) continue; 
        
        free(visited);

        if (count_turns(path) < 3) {
            free(path->points);
            free(path);
            continue; 
        }

        return path; 
    }
}

// 🌟 ✅ 修正 2：重構 Python 呼叫接口，改用指針安全寫入，保證起終點路徑 100% 同步
__declspec(dllexport) void generate_map_data(int* out_map, int* out_start, int* out_end, int* out_path_x, int* out_path_y, int* out_path_count) {
    Path* path = generate_strict_path();
    
    // 1. 清空傳進來的地圖陣列
    memset(out_map, 0, sizeof(int) * MAP_SIZE * MAP_SIZE);
    
    // 2. 填入起點與終點座標
    out_start[0] = path->points[0].x;
    out_start[1] = path->points[0].y;
    out_end[0] = path->points[path->count - 1].x;
    out_end[1] = path->points[path->count - 1].y;
    
    // 3. 填入整條 full_path 的點與長度，並渲染二維地圖
    *out_path_count = path->count;
    for (int i = 0; i < path->count; i++) {
        int x = path->points[i].x;
        int y = path->points[i].y;
        
        out_map[y * MAP_SIZE + x] = 1; // 畫二維通路
        out_path_x[i] = x;             // 紀錄路徑順序
        out_path_y[i] = y;
    }
    
    // 4. 安全釋放 C 內部暫存
    free(path->points);
    free(path);
}